bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/GarbageCollectorThread.java

```diff
@@ -264,12 +278,53 @@ public void safeRun() {
             // collection cycle started
             forceGarbageCollection.set(false);
         }
+
+    }
+
+    /**
+     * increase the threshold when the system in low load status.
+     */
+    private void activateMedianThreshold(){
+        majorCompactionThreshold = medianMajorCompactionThreshold;
+    }
+    /**
+     * increase the threshold when the system in a very low load status.
+     */
+    private void activateHighThreshold(){
+        majorCompactionThreshold = highMajorCompactionThreshold;
+    }
+
+    /**
+     * Check whether the configured cron is met and activate threshold.
+     */
+    private void changeMajorCompactionThreshold(){
```

Comment:
This strikes me as quite brittle.  It seems to me that you'll end up selecting the median (or high) threshhold several times in a row (approximately majorCompactionInterval / gcWaitTIme times).  It sort of works out because at line 350 all but 1 (or possibly 0 or 2, I'm not sure that the time between calls to now() couldn't cause a race condition where you skip a high/median compaction threshold or do it twice) will fail the curTime - lastMajorCompactionTime > majorCompactionInterval test.

It would be considerably simpler if you changed changeMajorCompactionThreshold to return the current threshold, move the call under the test at line 350, and maintain two pieces of state recording the last cron time where you returned a high or median value.